driver-core/src/main/com/mongodb/internal/async/SingleResultCallback.java

```diff
@@ -34,4 +36,32 @@ public interface SingleResultCallback<T> {
      * @throws Error Never, on the best effort basis.
      */
     void onResult(@Nullable T result, @Nullable Throwable t);
+
+    /**
+     * @return this callback as a handler
+     */
+    default AsyncCompletionHandler<T> asHandler() {
+        return new AsyncCompletionHandler<T>() {
+            @Override
+            public void completed(@Nullable final T result) {
+                onResult(result, null);
+            }
+            @Override
+            public void failed(final Throwable t) {
+                onResult(null, t);
+            }
+        };
+    }
+
+    default void complete(final SingleResultCallback<Void> callback) {
+        // takes a void callback (itself) to help ensure that this method
+        // is not accidentally used when "complete(T)" should have been used
+        // instead, since results are not marked nullable.
+        Assertions.assertTrue(callback == this);
+        this.onResult(null, null);
+    }
+
+    default void complete(final T result) {
```

Comment:
It seems cleaner to `throw`, since there is no chance that an ensuing `return` will be forgotten.

(This method would not be needed, except in some cases, there can be a non-async branch that needs to complete using the callback.)

Reply:
I was thinking that code like this in `AsyncRunnable`:

```
            } catch (Throwable t) {
                callback.onResult(null, t);
                return;
            }
```

would be cleaner as

```
            } catch (Throwable t) {
                callback.completeExceptionally(t);
                return;
            }
```

Reply:
This is true, but I think having it a less clean internally is better than exposing that second option externally, when it seems that in all cases we should prefer to write:
```
            } catch (Throwable t) {
                throw t;
            }
```
We can't do this internally in `AsyncRunnable`, because that is the place where thrown exceptions are caught and handled. And we can't have `completeExceptionally` was a private method on an interface, to prevent it being exposed. (Perhaps there is an external case where we cannot use `throw`, or where this is not cleaner?)

Reply:
I see your point for code that has been written using the new framework, but it still seems like a good thing to add for SingleResultCallback-based code where we're not quite ready to re-write using the new framework.  Everywhere that we currently have:

```
(result, t) -> { 
           if (t != null) {
                callback.onResult(null, t);
            } else {
               ....
               callback.onResult(newVal, null);
            }
```

would be clearer as

```
(result, t) -> { 
           if (t != null) {
                callback.completeExceptionally(t);
            } else {
               ....
               callback.complete(newVal);
            }
```

If we had been able to use default methods back when `SingleResultCallack` was first introduced (I think we still supported Java 6 back then?) we probably would have added then straightaway.

I'll leave this thread open for others to weigh in on.

Reply:
Fair enough, this sounds like a good idea for cases where we plan to use this outside of the API. Added.